File: packages/local_auth/local_auth_android/android/src/test/java/io/flutter/plugins/localauth/LocalAuthTest.java
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.localauth;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import android.app.Activity;
import android.app.KeyguardManager;
import android.app.NativeActivity;
import android.content.Context;
import androidx.biometric.BiometricManager;
import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.Lifecycle;
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.embedding.engine.plugins.lifecycle.HiddenLifecycleReference;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugins.localauth.AuthenticationHelper.AuthCompletionHandler;
import io.flutter.plugins.localauth.Messages.AuthClassification;
import io.flutter.plugins.localauth.Messages.AuthOptions;
import io.flutter.plugins.localauth.Messages.AuthResult;
import io.flutter.plugins.localauth.Messages.AuthStrings;
import io.flutter.plugins.localauth.Messages.Result;
import java.util.List;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

@RunWith(RobolectricTestRunner.class)
public class LocalAuthTest {
  static final AuthStrings dummyStrings =
      new AuthStrings.Builder()
          .setReason("a reason")
          .setBiometricHint("a hint")
          .setBiometricNotRecognized("biometric not recognized")
          .setBiometricRequiredTitle("biometric required")
          .setCancelButton("cancel")
          .setDeviceCredentialsRequiredTitle("credentials required")
          .setDeviceCredentialsSetupDescription("credentials setup description")
          .setGoToSettingsButton("go")
          .setGoToSettingsDescription("go to settings description")
          .setSignInTitle("sign in")
          .build();

  static final AuthOptions defaultOptions =
      new AuthOptions.Builder()
          .setBiometricOnly(false)
          .setSensitiveTransaction(false)
          .setSticky(false)
          .setUseErrorDialgs(false)
          .build();

  @Test
  public void authenticate_returnsErrorWhenAuthInProgress() {
    final LocalAuthPlugin plugin = new LocalAuthPlugin();
    plugin.authInProgress.set(true);
    @SuppressWarnings("unchecked")
    final Result<AuthResult> mockResult = mock(Result.class);
    plugin.authenticate(defaultOptions, dummyStrings, mockResult);
    ArgumentCaptor<AuthResult> captor = ArgumentCaptor.forClass(AuthResult.class);
    verify(mockResult).success(captor.capture());
    assertEquals(AuthResult.ERROR_ALREADY_IN_PROGRESS, captor.getValue());
  }

  @Test
  public void authenticate_returnsErrorWithNoForegroundActivity() {
    final LocalAuthPlugin plugin = new LocalAuthPlugin();
    @SuppressWarnings("unchecked")
    final Result<AuthResult> mockResult = mock(Result.class);

    plugin.authenticate(defaultOptions, dummyStrings, mockResult);
    ArgumentCaptor<AuthResult> captor = ArgumentCaptor.forClass(AuthResult.class);
    verify(mockResult).success(captor.capture());
    assertEquals(AuthResult.ERROR_NO_ACTIVITY, captor.getValue());
  }

  @Test
  public void authenticate_returnsErrorWhenActivityNotFragmentActivity() {
    final LocalAuthPlugin plugin = new LocalAuthPlugin();
    setPluginActivity(plugin, buildMockActivityWithContext(mock(NativeActivity.class)));
    @SuppressWarnings("unchecked")
    final Result<AuthResult> mockResult = mock(Result.class);
    plugin.authenticate(defaultOptions, dummyStrings, mockResult);
    ArgumentCaptor<AuthResult> captor = ArgumentCaptor.forClass(AuthResult.class);
    verify(mockResult).success(captor.capture());
    assertEquals(AuthResult.ERROR_NOT_FRAGMENT_ACTIVITY, captor.getValue());
  }

  @Test
  public void authenticate_returnsErrorWhenDeviceNotSupported() {
    final LocalAuthPlugin plugin = new LocalAuthPlugin();
    setPluginActivity(plugin, buildMockActivityWithContext(mock(FragmentActivity.class)));
    @SuppressWarnings("unchecked")
    final Result<AuthResult> mockResult = mock(Result.class);

    plugin.authenticate(defaultOptions, dummyStrings, mockResult);
    ArgumentCaptor<AuthResult> captor = ArgumentCaptor.forClass(AuthResult.class);
    verify(mockResult).success(captor.capture());
    assertEquals(AuthResult.ERROR_NOT_AVAILABLE, captor.getValue());
  }

  @Test
  public void authenticate_properlyConfiguresBiometricOnlyAuthenticationRequest() {
    final LocalAuthPlugin plugin = spy(new LocalAuthPlugin());
    setPluginActivity(plugin, buildMockActivityWithContext(mock(FragmentActivity.class)));
    when(plugin.isDeviceSupported()).thenReturn(true);

    final BiometricManager mockBiometricManager = mock(BiometricManager.class);
    when(mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK))
        .thenReturn(BiometricManager.BIOMETRIC_SUCCESS);
    when(mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.DEVICE_CREDENTIAL))
        .thenReturn(BiometricManager.BIOMETRIC_SUCCESS);
    plugin.setBiometricManager(mockBiometricManager);

    ArgumentCaptor<Boolean> allowCredentialsCaptor = ArgumentCaptor.forClass(Boolean.class);
    doNothing()
        .when(plugin)
        .sendAuthenticationRequest(
            any(AuthOptions.class),
            any(AuthStrings.class),
            allowCredentialsCaptor.capture(),
            any(AuthCompletionHandler.class));
    @SuppressWarnings("unchecked")
    final Result<AuthResult> mockResult = mock(Result.class);

    final AuthOptions options =
        new AuthOptions.Builder()
            .setBiometricOnly(true)
            .setSensitiveTransaction(false)
            .setSticky(false)
            .setUseErrorDialgs(false)
            .build();
    plugin.authenticate(options, dummyStrings, mockResult);
    assertFalse(allowCredentialsCaptor.getValue());
  }

  @Test
  @Config(sdk = 30)
  public void authenticate_properlyConfiguresBiometricAndDeviceCredentialAuthenticationRequest() {
    final LocalAuthPlugin plugin = spy(new LocalAuthPlugin());
    setPluginActivity(plugin, buildMockActivityWithContext(mock(FragmentActivity.class)));
    when(plugin.isDeviceSupported()).thenReturn(true);

    final BiometricManager mockBiometricManager = mock(BiometricManager.class);
    when(mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.DEVICE_CREDENTIAL))
        .thenReturn(BiometricManager.BIOMETRIC_SUCCESS);
    plugin.setBiometricManager(mockBiometricManager);

    ArgumentCaptor<Boolean> allowCredentialsCaptor = ArgumentCaptor.forClass(Boolean.class);
    doNothing()
        .when(plugin)
        .sendAuthenticationRequest(
            any(AuthOptions.class),
            any(AuthStrings.class),
            allowCredentialsCaptor.capture(),
            any(AuthCompletionHandler.class));
    @SuppressWarnings("unchecked")
    final Result<AuthResult> mockResult = mock(Result.class);

    plugin.authenticate(defaultOptions, dummyStrings, mockResult);
    assertTrue(allowCredentialsCaptor.getValue());
  }

  @Test
  @Config(sdk = 30)
  public void authenticate_properlyConfiguresDeviceCredentialOnlyAuthenticationRequest() {
    final LocalAuthPlugin plugin = spy(new LocalAuthPlugin());
    setPluginActivity(plugin, buildMockActivityWithContext(mock(FragmentActivity.class)));
    when(plugin.isDeviceSupported()).thenReturn(true);

    final BiometricManager mockBiometricManager = mock(BiometricManager.class);
    when(mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK))
        .thenReturn(BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED);
    when(mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.DEVICE_CREDENTIAL))
        .thenReturn(BiometricManager.BIOMETRIC_SUCCESS);
    plugin.setBiometricManager(mockBiometricManager);

    ArgumentCaptor<Boolean> allowCredentialsCaptor = ArgumentCaptor.forClass(Boolean.class);
    doNothing()
        .when(plugin)
        .sendAuthenticationRequest(
            any(AuthOptions.class),
            any(AuthStrings.class),
            allowCredentialsCaptor.capture(),
            any(AuthCompletionHandler.class));
    @SuppressWarnings("unchecked")
    final Result<AuthResult> mockResult = mock(Result.class);

    plugin.authenticate(defaultOptions, dummyStrings, mockResult);
    assertTrue(allowCredentialsCaptor.getValue());
  }

  @Test
  public void isDeviceSupportedReturnsFalse() {
    final LocalAuthPlugin plugin = new LocalAuthPlugin();
    assertFalse(plugin.isDeviceSupported());
  }

  @Test
  public void deviceCanSupportBiometrics_returnsTrueForPresentNonEnrolledBiometrics() {
    final LocalAuthPlugin plugin = new LocalAuthPlugin();
    final BiometricManager mockBiometricManager = mock(BiometricManager.class);
    when(mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK))
        .thenReturn(BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED);
    plugin.setBiometricManager(mockBiometricManager);

    assertTrue(plugin.deviceCanSupportBiometrics());
  }

  @Test
  public void deviceSupportsBiometrics_returnsTrueForPresentEnrolledBiometrics() {
    final LocalAuthPlugin plugin = new LocalAuthPlugin();
    final BiometricManager mockBiometricManager = mock(BiometricManager.class);
    when(mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK))
        .thenReturn(BiometricManager.BIOMETRIC_SUCCESS);
    plugin.setBiometricManager(mockBiometricManager);

    assertTrue(plugin.deviceCanSupportBiometrics());
  }

  @Test
  public void deviceSupportsBiometrics_returnsFalseForNoBiometricHardware() {
    final LocalAuthPlugin plugin = new LocalAuthPlugin();
    final BiometricManager mockBiometricManager = mock(BiometricManager.class);
    when(mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK))
        .thenReturn(BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE);
    plugin.setBiometricManager(mockBiometricManager);

    assertFalse(plugin.deviceCanSupportBiometrics());
  }

  @Test
  public void deviceSupportsBiometrics_returnsFalseForNullBiometricManager() {
    final LocalAuthPlugin plugin = new LocalAuthPlugin();
    plugin.setBiometricManager(null);

    assertFalse(plugin.deviceCanSupportBiometrics());
  }

  @Test
  public void onDetachedFromActivity_ShouldReleaseActivity() {
    final Activity mockActivity = mock(Activity.class);
    final ActivityPluginBinding mockActivityBinding = mock(ActivityPluginBinding.class);
    when(mockActivityBinding.getActivity()).thenReturn(mockActivity);

    Context mockContext = mock(Context.class);
    when(mockActivity.getBaseContext()).thenReturn(mockContext);
    when(mockActivity.getApplicationContext()).thenReturn(mockContext);

    final HiddenLifecycleReference mockLifecycleReference = mock(HiddenLifecycleReference.class);
    when(mockActivityBinding.getLifecycle()).thenReturn(mockLifecycleReference);

    final Lifecycle mockLifecycle = mock(Lifecycle.class);
    when(mockLifecycleReference.getLifecycle()).thenReturn(mockLifecycle);

    final FlutterPluginBinding mockPluginBinding = mock(FlutterPluginBinding.class);
    final BinaryMessenger mockMessenger = mock(BinaryMessenger.class);
    when(mockPluginBinding.getBinaryMessenger()).thenReturn(mockMessenger);

    final LocalAuthPlugin plugin = new LocalAuthPlugin();
    plugin.onAttachedToEngine(mockPluginBinding);
    plugin.onAttachedToActivity(mockActivityBinding);
    assertNotNull(plugin.getActivity());

    plugin.onDetachedFromActivity();
    assertNull(plugin.getActivity());
  }

  @Test
  public void getEnrolledBiometrics_shouldReturnEmptyList_withoutHardwarePresent() {
    final LocalAuthPlugin plugin = new LocalAuthPlugin();
    setPluginActivity(plugin, buildMockActivityWithContext(mock(Activity.class)));
    final BiometricManager mockBiometricManager = mock(BiometricManager.class);
    when(mockBiometricManager.canAuthenticate(anyInt()))
        .thenReturn(BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE);
    plugin.setBiometricManager(mockBiometricManager);

    final List<AuthClassification> enrolled = plugin.getEnrolledBiometrics();
    assertTrue(enrolled.isEmpty());
  }

  @Test
  public void getEnrolledBiometrics_shouldReturnEmptyList_withNoMethodsEnrolled() {
    final LocalAuthPlugin plugin = new LocalAuthPlugin();
    setPluginActivity(plugin, buildMockActivityWithContext(mock(Activity.class)));
    final BiometricManager mockBiometricManager = mock(BiometricManager.class);
    when(mockBiometricManager.canAuthenticate(anyInt()))
        .thenReturn(BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED);
    plugin.setBiometricManager(mockBiometricManager);

    final List<AuthClassification> enrolled = plugin.getEnrolledBiometrics();
    assertTrue(enrolled.isEmpty());
  }

  @Test
  public void getEnrolledBiometrics_shouldOnlyAddEnrolledBiometrics() {
    final LocalAuthPlugin plugin = new LocalAuthPlugin();
    setPluginActivity(plugin, buildMockActivityWithContext(mock(Activity.class)));
    final BiometricManager mockBiometricManager = mock(BiometricManager.class);
    when(mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK))
        .thenReturn(BiometricManager.BIOMETRIC_SUCCESS);
    when(mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_STRONG))
        .thenReturn(BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED);
    plugin.setBiometricManager(mockBiometricManager);

    final List<AuthClassification> enrolled = plugin.getEnrolledBiometrics();
    assertEquals(1, enrolled.size());
    assertEquals(AuthClassification.WEAK, enrolled.get(0));
  }

  @Test
  public void getEnrolledBiometrics_shouldAddStrongBiometrics() {
    final LocalAuthPlugin plugin = new LocalAuthPlugin();
    setPluginActivity(plugin, buildMockActivityWithContext(mock(Activity.class)));
    final BiometricManager mockBiometricManager = mock(BiometricManager.class);
    when(mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK))
        .thenReturn(BiometricManager.BIOMETRIC_SUCCESS);
    when(mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_STRONG))
        .thenReturn(BiometricManager.BIOMETRIC_SUCCESS);
    plugin.setBiometricManager(mockBiometricManager);

    final List<AuthClassification> enrolled = plugin.getEnrolledBiometrics();
    assertEquals(2, enrolled.size());
    assertEquals(AuthClassification.WEAK, enrolled.get(0));
    assertEquals(AuthClassification.STRONG, enrolled.get(1));
  }

  @Test
  @Config(sdk = 22)
  public void isDeviceSecure_returnsFalseOnBelowApi23() {
    final LocalAuthPlugin plugin = new LocalAuthPlugin();
    assertFalse(plugin.isDeviceSecure());
  }

  @Test
  @Config(sdk = 23)
  public void isDeviceSecure_returnsTrueIfDeviceIsSecure() {
    final LocalAuthPlugin plugin = new LocalAuthPlugin();
    KeyguardManager mockKeyguardManager = mock(KeyguardManager.class);
    plugin.setKeyguardManager(mockKeyguardManager);

    when(mockKeyguardManager.isDeviceSecure()).thenReturn(true);
    assertTrue(plugin.isDeviceSecure());

    when(mockKeyguardManager.isDeviceSecure()).thenReturn(false);
    assertFalse(plugin.isDeviceSecure());
  }

  @Test
  @Config(sdk = 30)
  public void
      canAuthenticateWithDeviceCredential_returnsTrueIfHasBiometricManagerSupportAboveApi30() {
    final LocalAuthPlugin plugin = new LocalAuthPlugin();
    final BiometricManager mockBiometricManager = mock(BiometricManager.class);
    plugin.setBiometricManager(mockBiometricManager);

    when(mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.DEVICE_CREDENTIAL))
        .thenReturn(BiometricManager.BIOMETRIC_SUCCESS);
    assertTrue(plugin.canAuthenticateWithDeviceCredential());

    when(mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.DEVICE_CREDENTIAL))
        .thenReturn(BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED);
    assertFalse(plugin.canAuthenticateWithDeviceCredential());
  }

  private Activity buildMockActivityWithContext(Activity mockActivity) {
    final Context mockContext = mock(Context.class);
    when(mockActivity.getBaseContext()).thenReturn(mockContext);
    when(mockActivity.getApplicationContext()).thenReturn(mockContext);
    return mockActivity;
  }

  private void setPluginActivity(LocalAuthPlugin plugin, Activity activity) {
    final HiddenLifecycleReference mockLifecycleReference = mock(HiddenLifecycleReference.class);
    final FlutterPluginBinding mockPluginBinding = mock(FlutterPluginBinding.class);
    final ActivityPluginBinding mockActivityBinding = mock(ActivityPluginBinding.class);
    final BinaryMessenger mockMessenger = mock(BinaryMessenger.class);
    when(mockPluginBinding.getBinaryMessenger()).thenReturn(mockMessenger);
    when(mockActivityBinding.getActivity()).thenReturn(activity);
    when(mockActivityBinding.getLifecycle()).thenReturn(mockLifecycleReference);
    plugin.onAttachedToEngine(mockPluginBinding);
    plugin.onAttachedToActivity(mockActivityBinding);
  }
}

File: packages/local_auth/local_auth_android/android/src/test/kotlin/io/flutter/plugins/localauth/LocalAuthTest.kt
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
package io.flutter.plugins.localauth

import android.app.Activity
import android.app.KeyguardManager
import android.app.NativeActivity
import android.content.Context
import androidx.biometric.BiometricManager
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.Lifecycle
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.embedding.engine.plugins.lifecycle.HiddenLifecycleReference
import io.flutter.plugin.common.BinaryMessenger
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentCaptor
import org.mockito.Mockito
import org.mockito.kotlin.any
import org.mockito.kotlin.eq
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

@RunWith(RobolectricTestRunner::class)
class LocalAuthTest {
  @Test
  fun authenticate_returnsErrorWhenAuthInProgress() {
    val plugin = LocalAuthPlugin()
    plugin.authInProgress.set(true)
    val callbackCalled = ArrayList<Boolean>()
    plugin.authenticate(defaultOptions, dummyStrings) { reply: Result<AuthResult> ->
      callbackCalled.add(true)
      Assert.assertEquals(AuthResultCode.ALREADY_IN_PROGRESS, reply.getOrNull()?.code)
    }
    Assert.assertTrue(callbackCalled[0])
  }

  @Test
  fun authenticate_returnsErrorWithNoForegroundActivity() {
    val plugin = LocalAuthPlugin()
    val callbackCalled = ArrayList<Boolean>()

    plugin.authenticate(defaultOptions, dummyStrings) { reply: Result<AuthResult> ->
      callbackCalled.add(true)
      Assert.assertEquals(AuthResultCode.NO_ACTIVITY, reply.getOrNull()?.code)
    }
    Assert.assertTrue(callbackCalled[0])
  }

  @Test
  fun authenticate_returnsErrorWhenActivityNotFragmentActivity() {
    val plugin = LocalAuthPlugin()
    setPluginActivity(
        plugin, buildMockActivityWithContext(Mockito.mock(NativeActivity::class.java)))
    val callbackCalled = ArrayList<Boolean>()
    plugin.authenticate(defaultOptions, dummyStrings) { reply: Result<AuthResult> ->
      callbackCalled.add(true)
      Assert.assertEquals(AuthResultCode.NOT_FRAGMENT_ACTIVITY, reply.getOrNull()?.code)
    }
    Assert.assertTrue(callbackCalled[0])
  }

  @Test
  fun authenticate_returnsErrorWhenDeviceNotSupported() {
    val plugin = LocalAuthPlugin()
    setPluginActivity(
        plugin, buildMockActivityWithContext(Mockito.mock(FragmentActivity::class.java)))
    val callbackCalled = ArrayList<Boolean>()

    plugin.authenticate(defaultOptions, dummyStrings) { reply: Result<AuthResult> ->
      callbackCalled.add(true)
      Assert.assertEquals(AuthResultCode.NO_CREDENTIALS, reply.getOrNull()?.code)
    }
    Assert.assertTrue(callbackCalled[0])
  }

  @Test
  fun authenticate_properlyConfiguresBiometricOnlyAuthenticationRequest() {
    val plugin = Mockito.spy(LocalAuthPlugin())
    val activity =
        buildMockActivityWithContext(Mockito.mock(FragmentActivity::class.java)) as FragmentActivity
    setPluginActivity(plugin, activity)
    Mockito.`when`<Boolean?>(plugin.isDeviceSupported()).thenReturn(true)

    val mockBiometricManager = Mockito.mock(BiometricManager::class.java)
    Mockito.`when`<Int?>(
            mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK))
        .thenReturn(BiometricManager.BIOMETRIC_SUCCESS)
    Mockito.`when`<Int?>(
            mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.DEVICE_CREDENTIAL))
        .thenReturn(BiometricManager.BIOMETRIC_SUCCESS)
    plugin.setBiometricManager(mockBiometricManager)

    val allowCredentialsCaptor = ArgumentCaptor.forClass(Boolean::class.java)
    Mockito.doNothing()
        .`when`(plugin)
        .sendAuthenticationRequest(
            any(), any(), allowCredentialsCaptor.capture(), eq(activity), any())
    val options = AuthOptions(biometricOnly = true, sensitiveTransaction = false, sticky = false)

    plugin.authenticate(options, dummyStrings) {}
    Assert.assertFalse(allowCredentialsCaptor.getValue())
  }

  @Test
  @Config(sdk = [30])
  fun authenticate_properlyConfiguresBiometricAndDeviceCredentialAuthenticationRequest() {
    val plugin = Mockito.spy(LocalAuthPlugin())
    val activity =
        buildMockActivityWithContext(Mockito.mock(FragmentActivity::class.java)) as FragmentActivity
    setPluginActivity(plugin, activity)
    Mockito.`when`<Boolean?>(plugin.isDeviceSupported()).thenReturn(true)

    val mockBiometricManager = Mockito.mock(BiometricManager::class.java)
    Mockito.`when`<Int?>(
            mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.DEVICE_CREDENTIAL))
        .thenReturn(BiometricManager.BIOMETRIC_SUCCESS)
    plugin.setBiometricManager(mockBiometricManager)

    val allowCredentialsCaptor = ArgumentCaptor.forClass(Boolean::class.java)
    Mockito.doNothing()
        .`when`(plugin)
        .sendAuthenticationRequest(
            any(), any(), allowCredentialsCaptor.capture()!!, eq(activity), any())
    plugin.authenticate(defaultOptions, dummyStrings) {}
    Assert.assertTrue(allowCredentialsCaptor.getValue())
  }

  @Test
  @Config(sdk = [30])
  fun authenticate_properlyConfiguresDeviceCredentialOnlyAuthenticationRequest() {
    val plugin = Mockito.spy(LocalAuthPlugin())
    val activity =
        buildMockActivityWithContext(Mockito.mock(FragmentActivity::class.java)) as FragmentActivity
    setPluginActivity(plugin, activity)
    Mockito.`when`<Boolean?>(plugin.isDeviceSupported()).thenReturn(true)

    val mockBiometricManager = Mockito.mock(BiometricManager::class.java)
    Mockito.`when`<Int?>(
            mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK))
        .thenReturn(BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED)
    Mockito.`when`<Int?>(
            mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.DEVICE_CREDENTIAL))
        .thenReturn(BiometricManager.BIOMETRIC_SUCCESS)
    plugin.setBiometricManager(mockBiometricManager)

    val allowCredentialsCaptor = ArgumentCaptor.forClass(Boolean::class.java)
    Mockito.doNothing()
        .`when`(plugin)
        .sendAuthenticationRequest(
            any(), any(), allowCredentialsCaptor.capture()!!, eq(activity), any())
    plugin.authenticate(defaultOptions, dummyStrings) {}
    Assert.assertTrue(allowCredentialsCaptor.getValue())
  }

  @Test
  fun isDeviceSupportedReturnsFalse() {
    val plugin = LocalAuthPlugin()
    Assert.assertFalse(plugin.isDeviceSupported())
  }

  @Test
  fun deviceCanSupportBiometrics_returnsTrueForPresentNonEnrolledBiometrics() {
    val plugin = LocalAuthPlugin()
    val mockBiometricManager = Mockito.mock(BiometricManager::class.java)
    Mockito.`when`<Int?>(
            mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK))
        .thenReturn(BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED)
    plugin.setBiometricManager(mockBiometricManager)

    Assert.assertTrue(plugin.deviceCanSupportBiometrics())
  }

  @Test
  fun deviceSupportsBiometrics_returnsTrueForPresentEnrolledBiometrics() {
    val plugin = LocalAuthPlugin()
    val mockBiometricManager = Mockito.mock(BiometricManager::class.java)
    Mockito.`when`<Int?>(
            mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK))
        .thenReturn(BiometricManager.BIOMETRIC_SUCCESS)
    plugin.setBiometricManager(mockBiometricManager)

    Assert.assertTrue(plugin.deviceCanSupportBiometrics())
  }

  @Test
  fun deviceSupportsBiometrics_returnsFalseForNoBiometricHardware() {
    val plugin = LocalAuthPlugin()
    val mockBiometricManager = Mockito.mock(BiometricManager::class.java)
    Mockito.`when`<Int?>(
            mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK))
        .thenReturn(BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE)
    plugin.setBiometricManager(mockBiometricManager)

    Assert.assertFalse(plugin.deviceCanSupportBiometrics())
  }

  @Test
  fun deviceSupportsBiometrics_returnsFalseForNullBiometricManager() {
    val plugin = LocalAuthPlugin()
    plugin.setBiometricManager(null)

    Assert.assertFalse(plugin.deviceCanSupportBiometrics())
  }

  @Test
  fun onDetachedFromActivity_ShouldReleaseActivity() {
    val mockActivity = Mockito.mock(Activity::class.java)
    val mockActivityBinding = Mockito.mock(ActivityPluginBinding::class.java)
    Mockito.`when`<Activity?>(mockActivityBinding.activity).thenReturn(mockActivity)

    val mockContext = Mockito.mock(Context::class.java)
    Mockito.`when`<Context?>(mockActivity.baseContext).thenReturn(mockContext)
    Mockito.`when`<Context?>(mockActivity.applicationContext).thenReturn(mockContext)

    val mockLifecycleReference = Mockito.mock(HiddenLifecycleReference::class.java)
    Mockito.`when`<Any?>(mockActivityBinding.lifecycle).thenReturn(mockLifecycleReference)

    val mockLifecycle = Mockito.mock(Lifecycle::class.java)
    Mockito.`when`<Lifecycle?>(mockLifecycleReference.lifecycle).thenReturn(mockLifecycle)

    val mockPluginBinding = Mockito.mock(FlutterPluginBinding::class.java)
    val mockMessenger = Mockito.mock(BinaryMessenger::class.java)
    Mockito.`when`<BinaryMessenger?>(mockPluginBinding.binaryMessenger).thenReturn(mockMessenger)

    val plugin = LocalAuthPlugin()
    plugin.onAttachedToEngine(mockPluginBinding)
    plugin.onAttachedToActivity(mockActivityBinding)
    Assert.assertNotNull(plugin.activity)

    plugin.onDetachedFromActivity()
    Assert.assertNull(plugin.activity)
  }

  @Test
  fun getEnrolledBiometrics_shouldReturnNullForNoActivity() {
    val plugin = LocalAuthPlugin()

    val enrolled: List<AuthClassification>? = plugin.getEnrolledBiometrics()
    Assert.assertNull(enrolled)
  }

  @Test
  fun getEnrolledBiometrics_shouldReturnEmptyList_withoutHardwarePresent() {
    val plugin = LocalAuthPlugin()
    setPluginActivity(plugin, buildMockActivityWithContext(Mockito.mock(Activity::class.java)))
    val mockBiometricManager = Mockito.mock(BiometricManager::class.java)
    Mockito.`when`<Int?>(mockBiometricManager.canAuthenticate(any()))
        .thenReturn(BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE)
    plugin.setBiometricManager(mockBiometricManager)

    val enrolled: List<AuthClassification>? = plugin.getEnrolledBiometrics()
    Assert.assertTrue(enrolled!!.isEmpty())
  }

  @Test
  fun getEnrolledBiometrics_shouldReturnEmptyList_withNoMethodsEnrolled() {
    val plugin = LocalAuthPlugin()
    setPluginActivity(plugin, buildMockActivityWithContext(Mockito.mock(Activity::class.java)))
    val mockBiometricManager = Mockito.mock(BiometricManager::class.java)
    Mockito.`when`<Int?>(mockBiometricManager.canAuthenticate(any()))
        .thenReturn(BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED)
    plugin.setBiometricManager(mockBiometricManager)

    val enrolled: List<AuthClassification>? = plugin.getEnrolledBiometrics()
    Assert.assertTrue(enrolled!!.isEmpty())
  }

  @Test
  fun getEnrolledBiometrics_shouldOnlyAddEnrolledBiometrics() {
    val plugin = LocalAuthPlugin()
    setPluginActivity(plugin, buildMockActivityWithContext(Mockito.mock(Activity::class.java)))
    val mockBiometricManager = Mockito.mock(BiometricManager::class.java)
    Mockito.`when`<Int?>(
            mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK))
        .thenReturn(BiometricManager.BIOMETRIC_SUCCESS)
    Mockito.`when`<Int?>(
            mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_STRONG))
        .thenReturn(BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED)
    plugin.setBiometricManager(mockBiometricManager)

    val enrolled: List<AuthClassification>? = plugin.getEnrolledBiometrics()
    Assert.assertEquals(1, enrolled!!.size.toLong())
    Assert.assertEquals(AuthClassification.WEAK, enrolled[0])
  }

  @Test
  fun getEnrolledBiometrics_shouldAddStrongBiometrics() {
    val plugin = LocalAuthPlugin()
    setPluginActivity(plugin, buildMockActivityWithContext(Mockito.mock(Activity::class.java)))
    val mockBiometricManager = Mockito.mock(BiometricManager::class.java)
    Mockito.`when`<Int?>(
            mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK))
        .thenReturn(BiometricManager.BIOMETRIC_SUCCESS)
    Mockito.`when`<Int?>(
            mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_STRONG))
        .thenReturn(BiometricManager.BIOMETRIC_SUCCESS)
    plugin.setBiometricManager(mockBiometricManager)

    val enrolled: List<AuthClassification>? = plugin.getEnrolledBiometrics()
    Assert.assertEquals(2, enrolled!!.size.toLong())
    Assert.assertEquals(AuthClassification.WEAK, enrolled[0])
    Assert.assertEquals(AuthClassification.STRONG, enrolled[1])
  }

  @Test
  fun isDeviceSecure_returnsTrueIfDeviceIsSecure() {
    val plugin = LocalAuthPlugin()
    val mockKeyguardManager = Mockito.mock(KeyguardManager::class.java)
    plugin.setKeyguardManager(mockKeyguardManager)

    Mockito.`when`<Boolean?>(mockKeyguardManager.isDeviceSecure).thenReturn(true)
    Assert.assertTrue(plugin.isDeviceSecure)

    Mockito.`when`<Boolean?>(mockKeyguardManager.isDeviceSecure).thenReturn(false)
    Assert.assertFalse(plugin.isDeviceSecure)
  }

  @Test
  @Config(sdk = [30])
  fun canAuthenticateWithDeviceCredential_returnsTrueIfHasBiometricManagerSupportAboveApi30() {
    val plugin = LocalAuthPlugin()
    val mockBiometricManager = Mockito.mock(BiometricManager::class.java)
    plugin.setBiometricManager(mockBiometricManager)

    Mockito.`when`<Int?>(
            mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.DEVICE_CREDENTIAL))
        .thenReturn(BiometricManager.BIOMETRIC_SUCCESS)
    Assert.assertTrue(plugin.canAuthenticateWithDeviceCredential())

    Mockito.`when`<Int?>(
            mockBiometricManager.canAuthenticate(BiometricManager.Authenticators.DEVICE_CREDENTIAL))
        .thenReturn(BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED)
    Assert.assertFalse(plugin.canAuthenticateWithDeviceCredential())
  }

  private fun buildMockActivityWithContext(mockActivity: Activity): Activity {
    val mockContext = Mockito.mock(Context::class.java)
    Mockito.`when`<Context?>(mockActivity.baseContext).thenReturn(mockContext)
    Mockito.`when`<Context?>(mockActivity.applicationContext).thenReturn(mockContext)
    return mockActivity
  }

  private fun setPluginActivity(plugin: LocalAuthPlugin, activity: Activity?) {
    val mockLifecycleReference = Mockito.mock(HiddenLifecycleReference::class.java)
    val mockPluginBinding = Mockito.mock(FlutterPluginBinding::class.java)
    val mockActivityBinding = Mockito.mock(ActivityPluginBinding::class.java)
    val mockMessenger = Mockito.mock(BinaryMessenger::class.java)
    Mockito.`when`<BinaryMessenger?>(mockPluginBinding.binaryMessenger).thenReturn(mockMessenger)
    Mockito.`when`<Activity?>(mockActivityBinding.activity).thenReturn(activity)
    Mockito.`when`<Any?>(mockActivityBinding.lifecycle).thenReturn(mockLifecycleReference)
    plugin.onAttachedToEngine(mockPluginBinding)
    plugin.onAttachedToActivity(mockActivityBinding)
  }

  companion object {
    val dummyStrings: AuthStrings = AuthStrings("a reason", "a hint", "cancel", "sign in")

    val defaultOptions: AuthOptions =
        AuthOptions(biometricOnly = false, sensitiveTransaction = false, sticky = false)
  }
}
